### PR TITLE
monk_1hotencoding & momentum 

### PR DESCRIPTION
**Pull request by gaetanoantonicchio**
Ho aggiunto labels, titoli ai grafici seguendo il layout presentato da Micheli nelle sue slides.
Ci sono due grafici separati, uno per accuracy, uno per loss.

### CODE REVIEW

File: src/monk_tests.py [hunk 1]
```diff
@@ -1,27 +1,51 @@
...
+        'weights_init': 'random'
+
     }
+
     model = Network(**parameters)
-    model.compile(opt='sgd', loss='squared', lrn_rate=0.5)
-    model.fit(inp=monk1_train, target=labels, epochs=50, batch_size=len(monk1_train))
+    model.print_net()
+    model.compile(opt='sgd', loss='squared', metr='class_acc', lrn_rate=0.6)
+    model.fit(inp=monk1_train, target=labels, epochs=80, batch_size=40)
```
Comment (gaetanoantonicchio) [inline src/monk_tests.py:51]: works only with batch_size = 1

File: src/network.py [hunk 2]
```diff
@@ -265,10 +265,11 @@ def __init__(self, input_dim=3, units_per_layer=(3, 2), acts=('relu', 'sigmoid')
 
             # for every unit in the current layer create layer's units
             for j in range(units_per_layer[i]):
+                #TODO: this part needs to be checked. It overrites the values assigned by weight_init
                 units.append(
-                    Unit(w=weights_inits(type=weights_init, n_weights=n_weights, lower_lim=0., upper_lim=1.,
+                    Unit(w=weights_inits(type=weights_init, n_weights=n_weights, lower_lim=0.1, upper_lim=0.2,
                                          value=weights_value),
-                         b=weights_inits(type=weights_init, n_weights=1, lower_lim=0., upper_lim=1.,
+                         b=weights_inits(type=weights_init, n_weights=1, lower_lim=0., upper_lim=0.01,
```
Comment (gaetanoantonicchio) [inline src/network.py:272]: Questa parte qui andava a sovrascrivere i valori che passavamo con weight_init tramite Network.
Per testare valori in range 0.1 - 0.2 (come suggerisce Micheli nelle slides sul progetto) ho dovuto cambiare i valori direttamente da qui.
Un altra roba, il bias veniva settato con valori tra 0 e 1. Quando il bias ha valori alti (>0.5) questo influisce molto sul training, ed i risultati sono un pò instabili. Secondo me conviene settare il bias a 0 oppure con valori vicini a 0.
Impostandolo manualmente                        
```suggestion
                         b=weights_inits(type=weights_init, n_weights=1, lower_lim=0., upper_lim=0.01,
```
i risultati sul training sono nettamente migliori. 

Un altra cosa, sulle slides di Micheli dice avoid:
- all zeros
- high values
- all equal valus 
A questo punto un inizializzazione uniform sembra inutile. Che dici?

File: src/optimizers.py [hunk 3]
```diff
@@ -68,22 +75,28 @@ def optimize(self, train_set, targets, epochs, batch_size=1):
...
                 grad_net = net.get_struct()
 
                 # cycle through patterns and targets within a batch
                 for pattern, target in zip(train_batch, targets_batch):
                     net_outputs = net.forward(inp=pattern)
-                    epoch_error += self.loss.func(predicted=net_outputs, target=target)
+                    epoch_error[:] += self.loss.func(predicted=net_outputs, target=target)
+                    epoch_metric[:] += self.metr.func(predicted=net_outputs, target=target)
+
+                    # TODO: DEBUG --> printing predicted and target for each forward pass
```
Comment (gaetanoantonicchio) [inline src/optimizers.py:95]: Ho aggiunto questa riga per controllare che ad ogni forward pass l' output predicted si avvicini sempre di più al target.

File: src/monk_tests.py [hunk 4]
```diff
@@ -1,27 +1,51 @@
...
 
     parameters = {
-        'input_dim': 6,
-        'units_per_layer': (3, 1),
-        'acts': ('relu', 'sigmoid'),
-        'weights_init': 'uniform',
-        'weights_value': 0.1
+        'input_dim': 17,
+        'units_per_layer': (5, 1),
+        'acts': ('sigmoid', 'sigmoid'),
```
Comment (gaetanoantonicchio) [inline src/monk_tests.py:43]: Testato anche con 'relu' nell'hidden layer e 'sigmoid' nell'output. 
I risultati sono nettamente migliori. La rete raggiunge 100% accuracy in maniera più smooth con meno epoche.
Ho modificato un pò i valori di eta. Infatti con eta grande la loss (pur decrescendo) era molto instabile. Poi ho dato un occhiata alle slides, e dice di usare un eta piccolo quando si uno sgd. Infatti con eta 0.2 i risultati sono migliori. 
loss
accuracy

File: src/optimizers.py [hunk 5]
```diff
@@ -96,13 +112,38 @@ def optimize(self, train_set, targets, epochs, batch_size=1):
...
                     grad_net.layers[i].biases /= float(batch_size)
-                    net.layers[i].weights += self.lrn_rate * grad_net.layers[i].weights
-                    net.layers[i].biases += self.lrn_rate * grad_net.layers[i].biases
-
-            epoch_error = np.sum(epoch_error) / len(epoch_error)
-            errors.append(epoch_error / float(len(train_set)))
-
+                    # momentum update
+                    #TODO:we need to assign customizable variable alpha instead of single value (Alex's help required)
+                    momentum_net.layers[i].weights = grad_net.layers[i].weights * 0.4
```
Comment (gaetanoantonicchio) [inline src/optimizers.py:117]: Il momentum sembra funzionare bene. Ovviamente così è brutto da vedere, bisogna assegnare una variabile alpha.
learning_curve_momentum
accuracy_momentum
